CommandBox 3.9.0

bdw429s opened the pull request:
CommandBox 3.9.0

ilovezfs commented:
it's passing CI with Java 9. What's the problem?

ilovezfs commented on Formula/commandbox.rb:
```diff
@@ -1,16 +1,16 @@
   bottle :unneeded
 
-  depends_on :java => "1.7+"
+  depends_on :java => "1.7+ < 9"
```
you'd need to require Java 8 specifically and use a wrapper script

bdw429s replied:
@ilovezfs Can you explain what a "wrapper script" means?  I just need to allow either Java 7 or 8, but not 9.  I looked and looked on Google but couldn't find the correct syntax for this.  I couldn't even locate the Homebrew docs that covered the valid expressions above.  I just guessed with what I typed and it seemed to make the builds pass.

bdw429s commented:
> it's passing CI with Java 9. 
@ilovezfs Not from what I saw.  Check the builds in the first commit (which failed)  My program is not compatible with Java 9.  I couldn't find the valid syntax to only allow java 7 and 8 in the docs.  I need CommandBox to not run on Java 9 or it's going to error out right now.

ilovezfs commented:
Ah, it passed because an unsatisfied requirement is an automatic pass.
```
==> SKIPPING: commandbox
Java 1.7+ < 9 is required to install this formula.
JavaRequirement unsatisfied!
You can install with Homebrew-Cask:
 brew cask install caskroom/versions/java8
You can download from:
 https://www.oracle.com/technetwork/java/javase/downloads/index.html
```

ilovezfs commented:
```
 I couldn't find the valid syntax to only allow java 7 and 8 in the docs.
```
depends_on :java => "x"
```
passes `x` to `/usr/libexec/java_home -v`

```
iMac-TMP:~ joe$ ls -1 /Library/Java/JavaVirtualMachines/
jdk-9.0.1.jdk
jdk1.8.0_151.jdk
zulu-7.jdk
zulu1.7.0_154.jdk
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.7+
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.8+
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 9.0+
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 9.0
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 8.0
Unable to find any JVMs matching version "8.0".
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.8
/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home
iMac-TMP:~ joe$ /usr/libexec/java_home -v 1.7
/Library/Java/JavaVirtualMachines/zulu-7.jdk/Contents/Home
iMac-TMP:~ joe$
```
So you need
```
depends_on :java => "1.8"
```

bdw429s commented:
> depends_on :java => "1.8"

@ilovezfs But CommandBox runs on Java 7 as well as I stated above.  Wouldn't the syntax you showed above only work on Java 8 or am I missing something?

ilovezfs commented:
You're not missing anything. To exclude Java 9 we have to pin to Java 8 specifically.

bdw429s commented:
> depends_on :java => "x"

:point_up:  That doesn't really tell me anything.  Is that a semantic versioning syntax?  Can you use greater than, less than?  Can you have a list of versions?  A range of versions?  All these questions and more were unanswered when I Googled and read through your docs.

bdw429s commented:
> To exclude Java 9 we have to pin to Java 8 specifically.

Hmm, that certainly feels wrong to say that my application "depends on Java 8" when it doesn't (Since java 7 is also valid) but I'll make that change.

ilovezfs commented:
`man java_home` the docs are terrible :)

ilovezfs commented:
In terms of the wrapper script.

```
    libexec.install "box"
    (bin/"box").write_env_script libexec/"box", Language::Java.java_home_env("1.8")
```

ilovezfs commented:
which results in
```
iMac-TMP:~ joe$ cat /usr/local/Cellar/commandbox/3.9.0/bin/box 
#!/bin/bash
JAVA_HOME="$(/usr/libexec/java_home --version 1.8)" exec "/usr/local/Cellar/commandbox/3.9.0/libexec/box" "$@"
iMac-TMP:~ joe$ 
```

ilovezfs commented:
@bdw429s as you can see it fails without the wrapper script.

jclausen commented:
@ilovezfs : Can you clarify on the way to implement the wrapper within the formula? I'm unable to find any documentation in the Formula cookbook.  For clarification on who's who here, @bdw429s is the developer of the binary and  I'm the original submitter of the formula.

bdw429s commented:
Thanks for the additional help @ilovezfs .  Like Jon, I'm still a little unsure of exactly what to do here.  
Am I just replacing these lines in my rb file:

```
  def install
    bin.install "box"
    doc.install resource("apidocs")
  end
```

with this?

```
  def install
   libexec.install "box"
    (bin/"box").write_env_script libexec/"box", Language::Java.java_home_env("1.8")
    doc.install resource("apidocs")
  end
```

ilovezfs commented:
>Am I just replacing these lines in my rb file:

yup.

bdw429s commented:
Thanks again for the help @ilovezfs Now that I added in the wrapper script (and fixed some indentation) it seems we've appeased the build gods.  Let me know if anything else needs done.

ilovezfs commented:
You're welcome :)

Shipped! Thanks for the upgrade @bdw429s! :rocket: 

And thanks for pitching in @jclausen!
